docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,391 @@
+[server]
+labels = "zone=<zone>,rack=<rack>,host=<host>"
+```
+
+Label constraints can be implemented by defining `label_constraints` field in placement rules. `+` and `-` correspond to property `op`. Specifically, `+` is equivalent to `In` and `-` is equivalent to `NotIn`.
+
+For example, `+zone=sh,+zone=bj` is equivalent to:
+
+```
+"label_constraints": [
+	{"key": "zone", "op": "in", "values": ["sh", “bj”]}
+]
+```
+
+### Role configuration
+
+`ROLE` in the statement defines the role of replicas. Newly added replicas are in such roles.
+
+`ROLE` is equivalent to property `role` in placement rules. it can be one of such values: leader, follower, learner, voter.
+
+`ROLE` must be specified in the statement.
+
+### Count configuration
+
+`COUNT` in the statement defines replicas count in current rule, it’s equivalent to property `count` in placement rules.
+
+`COUNT` must be specified in the statement.
+
+### Key range configuration
+
+Affected key range must also be defined in placement rules. Now that `table_name` is specified in the ALTER statement, key range can be inferred.
+
+Typically, key format is in such a format: `t_{table_id}_r_{pk_value}`, where `pk_value` may be `_tidb_rowid` in some cases. `table_id` can be inferred from `table_name`, thus key range can be `t_{table_id}_` to `t_{table_id+1}_`.
+
+### Database placement
+
+Defining placement of databases simplifies the procedures when there are very many tables. For example, in a multi-tenant scenarios, each user may have a single database. The dataset in one database is relatively small, and it’s rare to query toward multiple databases. In this case, a whole database can be placed in a single region to reduce multi-region latency.
+
+Placement of databases can be also defined through ALTER statements:
+
+```sql
+ALTER {DATABASE | SCHEMA} schema_name
+	{ADD | ALTER} PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
+```
+
+This statement indicates TiDB to add new replicas or alter placement for one database, including all tables in it.
+
+Since key range is not successive in one database, each table in the database should correspond to a placement rule, so there may be many placement rules.
+
+Creating or dropping a table should also affect the placement rules.
+
+### Partition placement
+
+Defining placement of partitions is useful for Geo-Partitioning. In the cases where data is very relevant to zones, Geo-Partitioning can be applied to reduce multi-region latency.
+
+In Geo-Partitioning, the table must be splitted into partitions, and each partition is placed in specific zones. There are 2 kinds of partition placement:
+
+* Place all replicas of one partition on one zone
+* Place only leaders of one partition on one zone
+
+It’s up to users to choose the right solution.
+
+Placement of partitions can be also defined through ALTER statements:
+
+```sql
+ALTER TABLE table_name ALTER PARTITION partition_name
+	{ADD | ALTER} PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
```

Comment:
How to make the leader and the majority of the followers of a partition to zone 1, other replicas to zone 2(to tolerate disaster on zone 1)?

If there are 3 replicas, can the following SQL achieve the goal:
```sql
alter table t alter partition p0
add placement label="+zone=1" role="leader" count=1,
add placement label="+zone=1" role="follower" count=1,
add placement label="+zone=2" role="follower" count=1;
```

Reply:
Good question.
As default placement rule is `add placement voter="" count=3`, the default rule must also be dropped.
So the SQL should be:
```sql
alter table t alter partition p0
	add placement label="+zone=1" role="leader",
	add placement label="+zone=1:1,+zone=2:1" role="follower",
	drop placement role=voter;
```

Reply:
If the placement rules are sent one by one, the PD may return error when `drop placement role=voter` is the first one. Because after executing `drop placement role=voter`, there would be no voters. 
So the rules should be sent all at once. What do you think? @xhebox